### PR TITLE
Capp 12029- add SSO support

### PR DESCRIPTION
**Pull request by michael-valershtein**

### CODE REVIEW

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 1]
```diff
@@ -225,14 +264,17 @@ public Response updateAliasPassword(@PathParam("alias") String alias, PasswordCo
...
-		// TODO - Find a way to validate current user password.
 		try {
 			AccessToken accessToken = BearerHelper.getTokenDataFromBearer(request).orNull();
 			if (accessToken != null && accessToken.getPreferredUsername().equals(alias)) {
 				ResponseBuilder response = Response.notModified();
 				if (passwordContainer.isDataValid()) {
-					aliasService.updateAliasePassword(alias, passwordContainer.getCurrentPassword(),
-							passwordContainer.getNewPassword());
+
+                    String issuer = accessToken.getIssuer();
```
Comment (dop54321) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java]: please fix indentaion

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 2]
```diff
@@ -52,23 +53,27 @@
...
 import org.keycloak.representations.AccessToken;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Controller;
+import org.springframework.util.CollectionUtils;
 
 import com.qmino.miredot.annotations.ReturnType;
 
+@SuppressWarnings("CheckStyle")
```
Comment (dop54321) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java]: why?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 3]
```diff
@@ -117,20 +123,43 @@ public Response crossOriginForInstallations(@Context HttpHeaders headers, @PathP
...
-	 *             TODO - Rename to registered
 	 */
 	@GET
 	@Path("/exists/{alias}")
 	@Produces(MediaType.APPLICATION_JSON)
 	@ReturnType("java.lang.Boolean")
-	public Response registered(@PathParam("alias") String alias, @Context HttpServletRequest request) {
-		if (aliasService.registered(alias))
-			return appendAllowOriginHeader(Response.ok().entity(Boolean.TRUE), request);
+	@Deprecated
```
Comment (dop54321) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:133]: Can you please add comment explaining why is this API deprecated and which API should the client use instead?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 4]
```diff
@@ -117,20 +123,43 @@ public Response crossOriginForInstallations(@Context HttpHeaders headers, @PathP
...
+	@GET
+	@Path("/registered/{alias}")
+	@Produces(MediaType.APPLICATION_JSON)
+	@ReturnType("java.lang.Boolean")
+	public Response registered(@PathParam("alias") String alias, @Context HttpServletRequest request) {
+		return isAliasRegistered(alias, request);
 	}
 
 	/**
+	 * @deprecated
```
Comment (dop54321) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:164]: Can you please add comment explaining why is this API deprecated and which API should the client use instead?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 5]
```diff
@@ -620,32 +662,28 @@ public Response deleteDestructive(@PathParam("id") String id, @Context HttpServl
...
-	@ReturnType("org.jboss.aerogear.unifiedpush.api.Alias")
-	public Response addClientScope(@Context HttpServletRequest request
-			, @QueryParam("clientScope") @DefaultValue(USER_TENANT_SCOPE) String clientScope) {
-		try {
-			int updated = aliasService.addClientScope(clientScope);
-			return appendAllowOriginHeader(Response.ok(Collections.singletonMap("count", updated)), request);
-		} catch (Exception e) {
-			logger.error("Cannot add client scope", e);
-			return appendAllowOriginHeader(Response.status(Status.INTERNAL_SERVER_ERROR), request);
+		//Todo should return error if more then one push applicationid for alias
```
Comment (dop54321) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java]: Is there a task for it? if so please add CAPP-*** to this comment

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java (hunk 4 above)
Comment (michael-valershtein) [reply jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:164]: I added the @deprecated to documentation only, the @Deprecated annotation was there long before me...

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java (hunk 3 above)
Comment (michael-valershtein) [reply jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:133]: DONE

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 6]
```diff
@@ -101,6 +105,9 @@ public Response crossOriginForInstallations(@Context HttpHeaders headers, @PathP
 	}
 
 	/**
+	 * @deprecated This endpoint path does not reflect what it does. Thus it will be replaced with "/registered/{alias}" as part of 13647 after
```
Comment (ilancb4) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java]: Please add CAPP- to 13647

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 7]
```diff
@@ -225,14 +265,17 @@ public Response updateAliasPassword(@PathParam("alias") String alias, PasswordCo
...
 			AccessToken accessToken = BearerHelper.getTokenDataFromBearer(request).orNull();
 			if (accessToken != null && accessToken.getPreferredUsername().equals(alias)) {
 				ResponseBuilder response = Response.notModified();
 				if (passwordContainer.isDataValid()) {
-					aliasService.updateAliasePassword(alias, passwordContainer.getCurrentPassword(),
-							passwordContainer.getNewPassword());
+
+					String issuer = accessToken.getIssuer();
+					String jwtApplicationName = URLUtils.getLastPart(issuer);
+
```
Comment (ilancb4) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:273]: Please document what is done here. 
the name `lastPart` is raising my concern. note the in production there are client names such as `c-test-dashboard-demo` which translate to the subdomain  c-test-dashboard-demo-mcs.c-b4.com

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 8]
```diff
@@ -620,32 +663,27 @@ public Response deleteDestructive(@PathParam("id") String id, @Context HttpServl
...
 		try {
-			int updated = aliasService.updateKCUsersGuids();
-			return appendAllowOriginHeader(Response.ok(Collections.singletonMap("count", updated)), request);
-		} catch (Exception e) {
-			logger.error("Cannot update guids", e);
-			return appendAllowOriginHeader(Response.status(Status.INTERNAL_SERVER_ERROR), request);
+			uri = new URI(request.getRequestURL().toString());
+		} catch (URISyntaxException e) {
+			logger.error("Unable to create URI from requestURI:" + request.getRequestURI(), e);
+			return appendAllowOriginHeader(Response.status(Status.BAD_REQUEST.getStatusCode(), INVALID_REQUEST_URI_RESPOONSE), request);
```
Comment (ilancb4) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:666]: this is the third time this code is repeated

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java [hunk 9]
```diff
@@ -620,32 +663,27 @@ public Response deleteDestructive(@PathParam("id") String id, @Context HttpServl
...
-			return appendAllowOriginHeader(Response.ok(Collections.singletonMap("count", updated)), request);
-		} catch (Exception e) {
-			logger.error("Cannot add client scope", e);
-			return appendAllowOriginHeader(Response.status(Status.INTERNAL_SERVER_ERROR), request);
+		Set<UserTenantInfo> tenantRelations = aliasService.getTenantRelations(alias);
+		if (CollectionUtils.isEmpty(tenantRelations)) {
+			return appendAllowOriginHeader(Response.ok().entity(Boolean.FALSE), request);
 		}
+
+		String pushApplicationID = tenantRelations.iterator().next().getPushId().toString();
```
Comment (ilancb4) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java]: ideally break chaining into multiple lines

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpoint.java [hunk 10]
```diff
@@ -54,4 +66,43 @@ public Response registerSecured(@DefaultValue("") @HeaderParam("x-ag-old-token")
...
+	 *
+	 * @statuscode 200 Successful storage of the alias.
+	 * @statuscode 404 if keyclaok user or UTR for the user not found
+	 *
+	 */
+	@POST
+	@Path("/syncUtr/{alias}")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	public Response syncUtr(@PathParam("alias") String alias, @Context HttpServletRequest request) {
```
Comment (ilancb4) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpoint.java]: Why do we need to receive `alias` here? 
Can't we get the alias of of the authenticated user info we get from the JWT passed through the bearer?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpoint.java [hunk 11]
```diff
@@ -54,4 +66,43 @@ public Response registerSecured(@DefaultValue("") @HeaderParam("x-ag-old-token")
...
+		}
+		keycloakService.updateTenantsExistingUser(alias, tenantRelations, app.getName());
+
+		String realmName = keycloakService.getRealmName(app.getName());
+
+		// add 'installation' realm role
+		keycloakService.addUserRealmRoles(Collections.singletonList(keycloakService.KEYCLOAK_ROLE_USER), alias, realmName);
+
+		// disable password
+		keycloakService.disableUserCredentials(alias, realmName);
```
Comment (ilancb4) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpoint.java:112]: consider changing the URL for this method as it does more then just syncing the UTR. 
perhaps rephrase to `bindWithSSO` or something which doesn't give the reader a false sense of what is actually going on

File: jaxrs/src/test/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpointTest.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+import com.datastax.driver.core.utils.UUIDs;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = { WebConfigTest.class })
+public class SecuredRegistrationEndpointTest extends RestEndpointTest {
+
+    @Inject
+    private IKeycloakService keycloakService;
+    @Inject
+    AliasService aliasService;
```
Comment (ilancb4) [inline jaxrs/src/test/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpointTest.java]: private?

File: jaxrs/src/test/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpointTest.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        String email = userName;
+        String other = "";
+
+        Alias alias = new Alias(UUID.fromString(app.getPushApplicationID()), id, email, other);
+        aliasDao.create(alias);
+
+        //invoke utrSync endpoint
+        String token = keycloakService.getUserAccessToken(userName, password, realmName, appName);
+        ResteasyClient client = new ResteasyClientBuilder().register(new Authenticator(app.getPushApplicationID(), app.getMasterSecret())).build();
+        ResteasyWebTarget target = client.target(getRestFullPath() + "/registry/type/syncUtr/" + "WRONG_" + alias.getEmail().toLowerCase());
```
Comment (ilancb4) [inline jaxrs/src/test/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/SecuredRegistrationEndpointTest.java]: test may become irrelevant if `alias` parameter is no longer accepted by API

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/PushApplicationDao.java [hunk 14]
```diff
@@ -88,7 +88,8 @@
      *
      * @param variantIDs list of variant ids
      *
-     * @return list of push applications
+     * @return list of push applications\
```
Comment (ilancb4) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/PushApplicationDao.java]: discard change

File: servers/src/main/java/org/jboss/aerogear/unifiedpush/auth/PathBasedKeycloakConfigResolver.java [hunk 15]
```diff
@@ -48,49 +52,55 @@
...
-				is = applicationContext.getResource("/WEB-INF/" + realm + ".json").getInputStream();
+				is = applicationContext.getResource("/WEB-INF/" + fileName + ".json").getInputStream();
 			} catch (IOException e) {
-				throw new IllegalStateException("Not able to find the file /" + realm + ".json");
+				throw new IllegalStateException("Not able to find the file /" + fileName + ".json");
 			}
 
 			if (is == null) {
-				throw new IllegalStateException("Not able to find the file /" + realm + ".json");
+				throw new IllegalStateException("Not able to find the file /" + fileName + ".json");
```
Comment (ilancb4) [inline servers/src/main/java/org/jboss/aerogear/unifiedpush/auth/PathBasedKeycloakConfigResolver.java]: please make the two messages slightly different so they'd be able to be differentiated in the log

File: servers/src/main/java/org/jboss/aerogear/unifiedpush/auth/SecurityConfig.java [hunk 16]
```diff
@@ -82,10 +82,11 @@ public void init(WebSecurity web) throws Exception {
 		web.ignoring().antMatchers("/rest/keycloak/config");
 		web.ignoring().antMatchers("/rest/heartbeat*");
 		web.ignoring().antMatchers("/rest/otp/*");
-		
-		// JWT validation internally 
+		web.ignoring().antMatchers("/rest/realms/**");
```
Comment (ilancb4) [inline servers/src/main/java/org/jboss/aerogear/unifiedpush/auth/SecurityConfig.java]: please remove as unused

File: service/src/main/java/org/jboss/aerogear/unifiedpush/service/impl/AliasServiceImpl.java [hunk 17]
```diff
@@ -180,9 +185,21 @@ public Associated associated(String alias, String fqdn) {
...
+					? pushApplicationService.findByPushApplicationID(aliasObj.getPushApplicationId().toString())
+					: pushApplication;
+			StringBuffer expectedHost = new StringBuffer();
+			String realmName = keycloakService.getRealmName(pushApplication.getName());
+			expectedHost.append(realmName)
+					.append(keycloakService.separator())
+					.append(conf.getRootUrlDomain());
+			if (host.equals(expectedHost.toString())
+					|| host.equals(conf.getRootUrlDomain())
+					|| realmName.equals(conf.getUpsiRealm())) {
```
Comment (ilancb4) [inline service/src/main/java/org/jboss/aerogear/unifiedpush/service/impl/AliasServiceImpl.java]: would be nice to document as it's kind of hard to read, but logic looks solid

**Review by ilancb4 — COMMENTED** [review]
looks good. Please fix and resubmit

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java (hunk 7 above)
Comment (michael-valershtein) [reply jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/AliasEndpoint.java:273]: This is not taken from hostname, but from the issuer url.

**Pull request merged**